### PR TITLE
A draft PR to review on-going changes for match_bool_prefix

### PR DESCRIPTION
**Pull request by MaxKsyunz**
### Description
Add support for match_bool_prefix function in SQL and PPL.
 
### Issues Resolved
#187
 
### Reviewers
@Bit-Quill/amazon-opensearch 

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/RelevanceQuery.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import org.opensearch.sql.expression.NamedArgumentExpression;
+import org.opensearch.sql.opensearch.storage.script.filter.lucene.LuceneQuery;
+
+/**
+ * Base class for query abstraction that builds a relevance query from function expression.
+ */
+public abstract class RelevanceQuery<T extends QueryBuilder> extends LuceneQuery {
+
+  protected Map<String, QueryBuilderStep<T>>
+      queryBuildActions;
```
Comment (acarbonetto) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/RelevanceQuery.java:25]: can you put this on the previous line?

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/RelevanceQuery.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+  public RelevanceQuery() {
+    super();
+    this.queryBuildActions = buildActionMap();
+  }
+
+  @Override
+  public QueryBuilder build(FunctionExpression func) {
+    Iterator<Expression> iterator = func.getArguments().iterator();
```
Comment (acarbonetto) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/RelevanceQuery.java:34]: please add the check to see if the number of arguments is < 2

Comment (acarbonetto) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/RelevanceQuery.java:34]: that being said, is there a common place to extract the field and query arguments?  Seems like we're doing this in a lot of places...

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/RelevanceQuery.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    NamedArgumentExpression query = (NamedArgumentExpression) iterator.next();
+    T queryBuilder = createQueryBuilder(field.getValue().valueOf(null).stringValue(),
+        query.getValue().valueOf(null).stringValue());
+    while (iterator.hasNext()) {
+      NamedArgumentExpression arg = (NamedArgumentExpression) iterator.next();
+      if (!queryBuildActions.containsKey(arg.getArgName())) {
+        throw new SemanticCheckException(String
+            .format("Parameter %s is invalid for match function.", arg.getArgName()));
+      }
+      (queryBuildActions
```
Comment (acarbonetto) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/RelevanceQuery.java:45]: why are there "()" around queryBuildAcions.get(...)?

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/RelevanceQuery.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * Base class for query abstraction that builds a relevance query from function expression.
+ */
+public abstract class RelevanceQuery<T extends QueryBuilder> extends LuceneQuery {
+
+  protected Map<String, QueryBuilderStep<T>>
+      queryBuildActions;
+
+  public RelevanceQuery() {
+    super();
```
Comment (acarbonetto) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/RelevanceQuery.java:28]: is this needed?

File: sql/src/test/java/org/opensearch/sql/sql/antlr/MatchBoolPrefixParserTest.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package org.opensearch.sql.sql.antlr;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.condition.EnabledIfEnvironmentVariable;
+import org.opensearch.sql.common.antlr.SyntaxParserTestBase;
```
Comment (acarbonetto) [inline sql/src/test/java/org/opensearch/sql/sql/antlr/MatchBoolPrefixParserTest.java:5]: is this used?

File: sql/src/test/java/org/opensearch/sql/sql/antlr/MatchBoolPrefixParserTest.java [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+@EnabledIfEnvironmentVariable(named = "INCLUDE_WIP", matches = ".*")
+public class MatchBoolPrefixParserTest extends SQLParserTest {
+
+  @Test
+  public void testDefaultParameters() {
+    acceptQuery("SELECT * FROM T WHERE MATCH_BOOL_PREFIX(message, 'query')");
+  }
+
+  @Test public void testOneParameter() {
```
Comment (acarbonetto) [inline sql/src/test/java/org/opensearch/sql/sql/antlr/MatchBoolPrefixParserTest.java:15]: put the annotation on a separate line please

File: sql/src/main/java/org/opensearch/sql/sql/antlr/SQLSyntaxParser.java [hunk 7]
```diff
@@ -16,13 +16,14 @@
 /**
  * SQL syntax parser which encapsulates an ANTLR parser.
  */
-public class SQLSyntaxParser {
+public class SQLSyntaxParser implements org.opensearch.sql.common.antlr.Parser {
```
[the remark this answers is not in the document]
Comment (MaxKsyunz) [reply sql/src/main/java/org/opensearch/sql/sql/antlr/SQLSyntaxParser.java:19]: I can. IntelliJ did that.

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/RelevanceQuery.java (hunk 2 above)
Comment (MaxKsyunz) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/RelevanceQuery.java:34]: That common code is in the match_phrase PR. I will migrate that block of code from there to here.

Once our various implementations start to integrate, RelevanceQuery will be the base class for single field queries.

I expect there will be a similar one for multi-field queries.

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/RelevanceQuery.java (hunk 3 above)
Comment (MaxKsyunz) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/RelevanceQuery.java:45]: 🤷 
probably leftover from it being cast.

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/RelevanceQuery.java (hunk 4 above)
Comment (MaxKsyunz) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/RelevanceQuery.java:28]: Don't think so, but IntelliJ put it there.

File: sql/src/test/java/org/opensearch/sql/sql/antlr/MatchBoolPrefixParserTest.java (hunk 5 above)
Comment (MaxKsyunz) [reply sql/src/test/java/org/opensearch/sql/sql/antlr/MatchBoolPrefixParserTest.java:5]: It used to be. 

If it's not, I'm surprised `checkStyleTest` didn't say anything about it.

**Pull request closed without merging**